### PR TITLE
WRP-3083: Increased code coverage for VirtualList

### PR DESCRIPTION
**Pull request by daniel-stoian-lgp**
### Checklist

* [x] I have read and understand the [contribution guide](http://enactjs.com/docs/developer-guide/contributing/)
* [ ] A [CHANGELOG entry](http://enactjs.com/docs/developer-guide/contributing/changelogs/) is included
* [ ] At least one test case is included for this feature or bug fix
* [x] Documentation was added or is not needed
* [x] Documentation was verified or is not changed
* [x] UI test was passed or is not needed
* [x] Screenshot test was verified or is not needed
* [ ] This is an API breaking change

### Issue Resolved / Feature Added
[//]: # (Describe the issue resolved or feature added by this pull request)
Increased code coverage for VirtualList

### Resolution
[//]: # (Does the code work as intended?)
[//]: # (What is the impact of this change and *why* was it made?)

### Additional Considerations
[//]: # (How should the change be tested?)
[//]: # (Are there any outstanding questions?)
[//]: # (Were any side-effects caused by the change?)
Please check the first comments of this PR about a scenario that isn't covered by the tests, but adding a test for it won't hit the missing lines

### Links
[//]: # (Related issues, references)
WRP-3083

### Comments

Enact-DCO-1.0-Signed-off-by: Daniel Stoian [daniel.stoian@lgepartner.com](mailto:daniel.stoian@lgepartner.com)

### CODE REVIEW

**Comment by daniel-stoian-lgp** [comment]
I also tried to add a test where focus was moved outside the VirtualList and returned to an item inside it in order to hit the function **calculatePositionOnFocus()** inisde useThemeVirtualList.js
Although testing this scenario live in a sample does hit the function, in the tests it is not hit.
In order to test, use the test below and also add a console.log at the beggining of the calculatePositionOnFocus() function

Test example:

	_test('should navigate focus outside the list from first item using arrow-up key', () => {
		render(
			<div>
				<Button>Hello Button</Button>
				<VirtualList
					clientSize={clientSize}
					dataSize={5}
					itemRenderer={renderItem}
					itemSize={itemSize}
					scrollMode="native"
				/>
			</div>
		);

		jest.useFakeTimers();
		const button = screen.getByRole('button');
		const list = screen.getByRole('list');
		const item0 = list.children.item(0).children.item(0);

		focus(item0);
		expect(currentFocusIndex).toBe(0);

		act(() => jest.advanceTimersByTime(1500));

		pressUpKey(item0);
		screen.debug()
		//expect(button).toHaveFocus();

		jest.useRealTimers();
	});_

File: VirtualList/tests/useEvent-specs.js [hunk 1]
```diff
@@ -0,0 +1,418 @@
...
+
+import {ImageItem} from '../../ImageItem';
+import Item from '../../Item';
+import VirtualList, {VirtualGridList} from '../VirtualList';
+
+const focus = (elm) => fireEvent.focus(elm);
+
+const keyDownUp = (keyCode) => (elm) => {
+	fireEvent.keyDown(elm, {keyCode});
+	return fireEvent.keyUp(elm, {keyCode});
```
Comment (alexandrumorariu) [inline VirtualList/tests/useEvent-specs.js:12]: This return statement seems unnecessary.

File: VirtualList/tests/VirtualGridList-native-specs.js [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+		svgGenerator;
+
+	beforeEach(() => {
+		clientSize = {clientWidth: 1280, clientHeight: 720};
+		dataSize = 200;
+		items = [];
+		itemSize = {minWidth: 300, minHeight:240};
+
+		renderItem = ({index, ...rest}) => { // eslint-disable-line enact/display-name
+			const {text, subText, source} = items[index];
```
Comment (alexandrumorariu) [inline VirtualList/tests/VirtualGridList-native-specs.js:23]: Order variables please.

File: VirtualList/tests/VirtualGridList-native-specs.js [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+	describe('Scrollbar visibility', () => {
+		test('should render both horizontal and vertical scrollbars when `horizontalScrollbar` and `verticalScrollbar` are `visible`', () => {
+			render(
+				<VirtualGridList
+					clientSize={clientSize}
+					dataSize={dataSize}
+					direction="horizontal"
+					itemRenderer={renderItem}
+					itemSize={itemSize}
+					horizontalScrollbar="visible"
```
Comment (alexandrumorariu) [inline VirtualList/tests/VirtualGridList-native-specs.js:166]: Order props

File: VirtualList/tests/VirtualGridList-specs.js [hunk 4]
```diff
@@ -0,0 +1,621 @@
...
+			onScrollStopCount++;
+			resultScrollLeft = e.scrollLeft;
+			resultScrollTop = e.scrollTop;
+
+			testCase();
+			done();
+		};
+
+		renderItem = ({index, ...rest}) => { // eslint-disable-line enact/display-name
+			const {text, subText, source} = items[index];
```
Comment (alexandrumorariu) [inline VirtualList/tests/VirtualGridList-specs.js:59]: Order props please.

File: VirtualList/tests/VirtualGridList-native-specs.js [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+					direction="horizontal"
+					itemRenderer={renderItem}
+					itemSize={itemSize}
+					horizontalScrollbar="hidden"
+					scrollMode="native"
+					verticalScrollbar="hidden"
+				/>
+			);
+
+			const VirtualGridListRoot =  screen.getByRole('list').parentElement.parentElement.parentElement.parentElement;
```
Comment (alexandrumorariu) [inline VirtualList/tests/VirtualGridList-native-specs.js:214]: I do not belive that capital letter is justified here.

File: VirtualList/tests/VirtualGridList-specs.js [hunk 6]
```diff
@@ -0,0 +1,621 @@
...
+			render(
+				<VirtualGridList
+					clientSize={clientSize}
+					dataSize={dataSize}
+					itemRenderer={renderItem}
+					itemSize={itemSize}
+				/>
+			);
+
+			const VirtualGridListRoot =  screen.getByRole('list').parentElement.parentElement.parentElement.parentElement;
```
Comment (alexandrumorariu) [inline VirtualList/tests/VirtualGridList-specs.js:231]: I do not belive that capital letter is justified here.

File: VirtualList/tests/VirtualGridList-specs.js [hunk 7]
```diff
@@ -0,0 +1,621 @@
...
+					dataSize={dataSize}
+					direction="horizontal"
+					horizontalScrollbar="hidden"
+					itemRenderer={renderItem}
+					itemSize={itemSize}
+					verticalScrollbar="hidden"
+				/>
+			);
+
+			const VirtualGridListRoot =  screen.getByRole('list').parentElement.parentElement.parentElement.parentElement;
```
Comment (alexandrumorariu) [inline VirtualList/tests/VirtualGridList-specs.js:253]: I do not belive that capital letter is justified here.

File: VirtualList/tests/VirtualGridList-specs.js [hunk 8]
```diff
@@ -0,0 +1,621 @@
...
+				expect(resultScrollTop).toBeGreaterThan(0);
+			});
+
+			render(
+				<VirtualGridList
+					clientSize={clientSize}
+					dataSize={dataSize}
+					itemRenderer={renderItem}
+					itemSize={itemSize}
+					onScrollStop={onScrollStop}
```
Comment (alexandrumorariu) [inline VirtualList/tests/VirtualGridList-specs.js:542]: Order props

File: VirtualList/tests/VirtualGridList-specs.js [hunk 9]
```diff
@@ -0,0 +1,621 @@
...
+			const fn = jest.fn();
+
+			render(
+				<VirtualGridList
+					clientSize={clientSize}
+					dataSize={dataSize}
+					itemRenderer={renderItem}
+					itemSize={itemSize}
+					noScrollByWheel
+					onScrollStop={fn}
```
Comment (alexandrumorariu) [inline VirtualList/tests/VirtualGridList-specs.js:565]: Order props

File: VirtualList/tests/VirtualList-specs.js [hunk 10]
```diff
@@ -257,8 +366,72 @@ describe('VirtualList', () => {
...
+				expect(resultScrollTop).toBeGreaterThan(0);
+			});
+
+			render(
+				<VirtualList
+					clientSize={clientSize}
+					dataSize={dataSize}
+					itemRenderer={renderItem}
+					itemSize={itemSize}
+					onScrollStop={onScrollStop}
```
Comment (alexandrumorariu) [inline VirtualList/tests/VirtualList-specs.js:394]: Order props

File: VirtualList/tests/VirtualList-specs.js [hunk 11]
```diff
@@ -257,8 +366,72 @@ describe('VirtualList', () => {
...
+			const fn = jest.fn();
+
+			render(
+				<VirtualList
+					clientSize={clientSize}
+					dataSize={dataSize}
+					itemRenderer={renderItem}
+					itemSize={itemSize}
+					noScrollByWheel
+					onScrollStop={fn}
```
Comment (alexandrumorariu) [inline VirtualList/tests/VirtualList-specs.js:417]: Order props

File: VirtualList/tests/useEvent-specs.js [hunk 12]
```diff
@@ -0,0 +1,418 @@
...
+		imageItemSize = {minWidth: 300, minHeight: 240};
+		items = [];
+		itemSize = 60;
+
+		handlerOnFocus = (index) => () => {
+			currentFocusIndex = index;
+		};
+
+		renderImageItem = ({index, ...rest}) => { // eslint-disable-line enact/display-name
+			const {name, subText, source} = items[index];
```
Comment (alexandrumorariu) [inline VirtualList/tests/useEvent-specs.js:47]: Order props

File: VirtualList/tests/useEvent-specs.js [hunk 13]
```diff
@@ -0,0 +1,418 @@
...
+
+			return (
+				<ImageItem {...rest} label={subText} src={source} onFocus={handlerOnFocus(index)}>
```
Comment (alexandrumorariu) [inline VirtualList/tests/useEvent-specs.js:50]: Order props

File: VirtualList/tests/useEvent-specs.js [hunk 14]
```diff
@@ -0,0 +1,418 @@
...
+		expect(currentFocusIndex).toBe(0);
+
+		jest.useRealTimers();
+	});
+
+	test('should not navigate focus using arrow-up/down key when `direction` is horizontal', () => {
+		render(
+			<VirtualList
+				direction="horizontal"
+				clientSize={clientSize}
```
Comment (alexandrumorariu) [inline VirtualList/tests/useEvent-specs.js:178]: Order props

File: VirtualList/tests/useEvent-specs.js [hunk 15]
```diff
@@ -0,0 +1,418 @@
...
+
+		pressUpKey(item0);
+		expect(currentFocusIndex).toBe(0);
+	});
+
+	test('should navigate focus using arrow-left/right key when `direction` is horizontal', () => {
+		render(
+			<VirtualList
+				direction="horizontal"
+				clientSize={clientSize}
```
Comment (alexandrumorariu) [inline VirtualList/tests/useEvent-specs.js:202]: Order props

File: VirtualList/tests/useEvent-specs.js [hunk 16]
```diff
@@ -0,0 +1,418 @@
...
+		expect(spy).toHaveBeenCalled();
+	});
+
+	describe('VirtualGridList useEvent', () => {
+		test('should navigate focus using arrow-up/down key', () => {
+			render(
+				<VirtualGridList
+					clientSize={clientSize}
+					direction="vertical"
+					dataSize={dataSize}
```
Comment (alexandrumorariu) [inline VirtualList/tests/useEvent-specs.js:286]: Order props

**Review by alexandrumorariu — APPROVED** [review]
LGTM

File: VirtualList/tests/VirtualGridList-specs.js [hunk 17]
```diff
@@ -0,0 +1,621 @@
...
+			/>
+		);
+
+		const expected = 'Item 000';
+		const actual = screen.getByRole('list').children.item(0).textContent;
+
+		expect(actual).toBe(expected);
+	});
+
+	test('cannot render items when `clientSize` and outer DOM size are not specified', () => {
```
Comment (adrian-cocoara-lgp) [inline VirtualList/tests/VirtualGridList-specs.js:124]:
```suggestion
	test('should not render items when `clientSize` and outer DOM size are not specified', () => {
```

File: VirtualList/tests/VirtualGridList-specs.js [hunk 18]
```diff
@@ -0,0 +1,621 @@
...
+					onScrollStop={onScrollStop}
+					spacing={newSpacing}
+				/>
+			);
+
+			act(() => myScrollTo({align: 'bottom', animate: false}));
+		});
+
+		describe('scroll events', () => {
+			test('should call onScrollStart once', () => {
```
Comment (adrian-cocoara-lgp) [inline VirtualList/tests/VirtualGridList-specs.js:456]:
```suggestion
			test('should call `onScrollStart` once', () => {
```

File: VirtualList/tests/VirtualGridList-specs.js [hunk 19]
```diff
@@ -0,0 +1,621 @@
...
+				);
+
+				act(() => myScrollTo({position: {y: 200}, animate: false}));
+
+				const expected = 1;
+
+				expect(onScrollStartCount).toBe(expected);
+			});
+
+			test('should call onScroll once', () => {
```
Comment (adrian-cocoara-lgp) [inline VirtualList/tests/VirtualGridList-specs.js:475]:
```suggestion
			test('should call `onScroll` once', () => {
```

File: VirtualList/tests/VirtualGridList-specs.js [hunk 20]
```diff
@@ -0,0 +1,621 @@
...
+				);
+
+				act(() => myScrollTo({position: {y: 200}, animate: false}));
+
+				const expected = 1;
+
+				expect(onScrollCount).toBe(expected);
+			});
+
+			test('should call onScrollStop once', (done) => {
```
Comment (adrian-cocoara-lgp) [inline VirtualList/tests/VirtualGridList-specs.js:494]:
```suggestion
			test('should call `onScrollStop` once', (done) => {
```

**Review by adrian-cocoara-lgp — APPROVED** [review]
LGTM

File: VirtualList/tests/VirtualList-specs.js [hunk 21]
```diff
@@ -282,7 +455,7 @@ describe('VirtualList', () => {
 				clientSize={clientSize}
 				dataSize={itemArray.length}
 				itemRenderer={renderItemArray}
-				itemSize={60}
+				itemSize={itemSize}
```
Comment (stanca-pop-lgp) [inline VirtualList/tests/VirtualList-specs.js:458]: Please leave empty row before rerender (line 454).

File: VirtualList/tests/useEvent-native-specs.js [hunk 22]
```diff
@@ -0,0 +1,261 @@
...
+			items.push({name: 'Account ' + i});
+		}
+	});
+
+	afterEach(() => {
+		clientSize = null;
+		dataSize = null;
+		items = [];
+		itemSize = 60;
+		items = null;
```
Comment (stanca-pop-lgp) [inline VirtualList/tests/useEvent-native-specs.js:59]: Is assigning null to items which was previously assigned an empty array intended?

File: VirtualList/tests/useEvent-native-specs.js [hunk 23]
```diff
@@ -0,0 +1,261 @@
...
+		const item10 = list.children.item(10).children.item(0);
+
+		focus(item9);
+		expect(currentFocusIndex).toBe(9);
+
+		pressDownKey(item9);
+		expect(currentFocusIndex).toBe(10);
+
+		pressPageDownKey(item10);
+
```
Comment (stanca-pop-lgp) [inline VirtualList/tests/useEvent-native-specs.js:256]: Remove empty line.

**Review by stanca-pop-lgp — CHANGES_REQUESTED** [review]

File: VirtualList/tests/VirtualGridList-native-specs.js [hunk 24]
```diff
@@ -0,0 +1,265 @@
...
+			/>
+		);
+
+		const expected = 20; // (4 * (360 / 240)) + 4 * 3
+		const actual = screen.getByRole('list').children.length;
+
+		expect(actual).toBe(expected);
+	});
+
+	test('should render only one scrollbar', () => {
```
[the remark this answers is not in the document]
Comment (daniel-stoian-lgp) [reply VirtualList/tests/VirtualGridList-native-specs.js:139]: i added more description. Regarding the expected result of 2, the structure of the component is the cause. Check also sandstone tests for the similar approach

**Review by stanca-pop-lgp — APPROVED** [review]
LGTM

**Pull request merged**